lib/Parser/Parse.cpp

```diff
@@ -4713,15 +4671,72 @@ bool Parser::ParseFncDeclHelper(ParseNodePtr pnodeFnc, ParseNodePtr pnodeFncPare
     RestorePoint beginNameHint;
     m_pscan->Capture(&beginNameHint);
 
-    ParseNodePtr pnodeFncExprScope = nullptr;
+    ParseNodePtr pnodeFncExprScope = nullptr;    
     Scope *fncExprScope = nullptr;
     if (!fDeclaration)
     {
         pnodeFncExprScope = StartParseBlock<buildAST>(PnodeBlockType::Function, ScopeType_FuncExpr);
         fncExprScope = pnodeFncExprScope->sxBlock.scope;
+
+        // Function expression: push the new function onto the stack now so that the name (if any) will be
+        // local to the new function.
+
+        if (buildAST)
+        {
+            // Make this the current function and start its sub-function list.
+            m_currentNodeFunc = pnodeFnc;
+
+            Assert(m_currentNodeDeferredFunc == nullptr);
+
+            if (!fLambda)
+            {
+                m_currentNodeNonLambdaFunc = pnodeFnc;
+            }
+        }
+        else // if !buildAST
+        {
+            AnalysisAssert(pnodeFnc);
+
+            if (!fLambda)
+            {
+                m_currentNodeNonLambdaDeferredFunc = pnodeFnc;
+            }
+
+            m_currentNodeDeferredFunc = pnodeFnc;
+        }
     }
 
-    *pHasName = !fLambda && this->ParseFncNames<buildAST>(pnodeFnc, pnodeFncParent, flags, &lastNodeRef);
+    *pHasName = !fLambda && this->ParseFncNames<buildAST>(pnodeFnc, pnodeFncSave, flags, &lastNodeRef);
+
+    if (fDeclaration)
+    {
+        // Declaration statement: push the new function now, after parsing the name, so the name is local to the
+        // enclosing function.
+
+        if (buildAST)
+        {
+            // Make this the current function and start its sub-function list.
+            m_currentNodeFunc = pnodeFnc;
+
+            Assert(m_currentNodeDeferredFunc == nullptr);
+
+            if (!fLambda)
+            {
+                m_currentNodeNonLambdaFunc = pnodeFnc;
+            }
+        }
+        else // if !buildAST
+        {
+            AnalysisAssert(pnodeFnc);
+
+            if (!fLambda)
+            {
+                m_currentNodeNonLambdaDeferredFunc = pnodeFnc;
+            }
+
+            m_currentNodeDeferredFunc = pnodeFnc;
+        }
```

Comment:
True. It was a little more painful to do that the way I originally structured the change, but it's easy now, so I'll do it.